File: src/src/lib/endpoints/checkpoints.ts
import type { Checkpoint } from "../types";
import { requestJson } from "../api-client";

// Checkpoint listing + rollback go through the Control_Plane HTTP API only
// (Requirement 20.4): the Frontend never reaches the Checkpoint_Manager or the
// kernel directly. These routes are defined in
// control-plane/cp/handlers/checkpoints.py:
//   GET  /api/v1/runs/{runId}/checkpoints          (Requirement 13.3)
//   POST /api/v1/checkpoints/{checkpointId}/rollback (Requirement 13.2)
const RUNS_BASE = "/api/v1/runs";
const CHECKPOINTS_BASE = "/api/v1/checkpoints";

/**
 * List the Checkpoints recorded for a Workflow_Run (Requirement 13.3).
 *
 * The Control_Plane returns them ordered by creation time for a stable,
 * reviewable listing. A run with no checkpoints yields an empty list.
 */
export async function listCheckpoints(runId: string): Promise<Checkpoint[]> {
  const payload = (await requestJson(
    `${RUNS_BASE}/${encodeURIComponent(runId)}/checkpoints`,
    "GET",
  )) as { checkpoints?: Checkpoint[]; runId?: string };
  return payload.checkpoints ?? [];
}

/**
 * Request that the Kernel restore the workspace to a Checkpoint's snapshot
 * (Requirement 13.2). Returns the checkpoint that was restored.
 */
export async function rollbackCheckpoint(
  checkpointId: string,
): Promise<Checkpoint> {
  const payload = (await requestJson(
    `${CHECKPOINTS_BASE}/${encodeURIComponent(checkpointId)}/rollback`,
    "POST",
  )) as { checkpoint: Checkpoint };
  return payload.checkpoint;
}

File: control-plane/cp/handlers/checkpoints.py
"""Checkpoint routes: list a run's checkpoints and request rollback.

These routes expose the Checkpoint_Manager (``core/checkpoint_manager.py``) over
HTTP:

* ``GET  /api/v1/runs/{run_id}/checkpoints`` lists the Checkpoints recorded for a
  Workflow_Run (Requirement 13.3).
* ``POST /api/v1/checkpoints/{checkpoint_id}/rollback`` requests that the Kernel
  restore the workspace to a Checkpoint's snapshot (Requirement 13.2).

The handlers reach the Checkpoint_Manager through ``runtime`` accessor methods
(``runtime.list_checkpoints`` / ``runtime.restore_checkpoint``); runtime wiring
itself is handled separately (task 30.10).
"""

from __future__ import annotations

from typing import Any
from urllib.parse import unquote

from cp.handlers.router import register
from cp.http_responders import write_json
from cp.http_utils import require_method, split_segments

# ---------------------------------------------------------------------------
# Listing  /api/v1/runs/{run_id}/checkpoints   (Requirement 13.3)
# ---------------------------------------------------------------------------


def _handle_run_checkpoints(
    handler: Any, method: str, _path: str, _query: Any, _body: Any, runtime: Any
) -> None:
    """GET /api/v1/runs/{run_id}/checkpoints"""
    require_method(method, {"GET"})
    segments = split_segments(_path)
    run_id = unquote(segments[3])
    write_json(
        handler,
        {
            "checkpoints": runtime.list_checkpoints(run_id),
            "runId": run_id,
        },
    )


# ---------------------------------------------------------------------------
# Rollback  /api/v1/checkpoints/{checkpoint_id}/rollback   (Requirement 13.2)
# ---------------------------------------------------------------------------


def _handle_checkpoint_rollback(
    handler: Any, method: str, _path: str, _query: Any, _body: Any, runtime: Any
) -> None:
    """POST /api/v1/checkpoints/{checkpoint_id}/rollback"""
    require_method(method, {"POST"})
    segments = split_segments(_path)
    checkpoint_id = unquote(segments[3])
    suffix = segments[4]
    if suffix != "rollback":
        raise KeyError(f"checkpoint action `{suffix}` is not supported")
    checkpoint = runtime.restore_checkpoint(checkpoint_id)
    write_json(handler, {"checkpoint": checkpoint})


# -- registration ------------------------------------------------------------

register("*", ("api", "v1", "runs", "*", "checkpoints"), _handle_run_checkpoints)
register(
    "*", ("api", "v1", "checkpoints", "*", "rollback"), _handle_checkpoint_rollback
)
